src/modules/fancyzones/lib/FancyZones.cpp

```diff
@@ -566,49 +575,138 @@ LRESULT CALLBACK FancyZones::s_WndProc(HWND window, UINT message, WPARAM wparam,
+          }
+          return std::nullopt;
+      }
+    };
+    auto callback = [](HMONITOR monitor, HDC, RECT *, LPARAM captured_ptr) -> BOOL
+    {
+        const auto doContinueEnumeration = TRUE;
         MONITORINFOEX mi;
         mi.cbSize = sizeof(mi);
-        if (GetMonitorInfo(monitor, &mi))
+        if (GetMonitorInfo(monitor, &mi) == FALSE)
         {
-            DISPLAY_DEVICE displayDevice = { sizeof(displayDevice) };
-            PCWSTR deviceId = nullptr;
+          return doContinueEnumeration;
+        }
+        DISPLAY_DEVICE displayDevice = { sizeof(displayDevice) };
 
-            bool validMonitor = true;
-            if (EnumDisplayDevices(mi.szDevice, 0, &displayDevice, 1))
+        if (EnumDisplayDevices(mi.szDevice, 0, &displayDevice, 1))
+        {
+            if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER))
             {
-                if (WI_IsFlagSet(displayDevice.StateFlags, DISPLAY_DEVICE_MIRRORING_DRIVER))
-                {
-                    validMonitor = FALSE;
-                }
-                else if (displayDevice.DeviceID[0] != L'\0')
-                {
-                    deviceId = displayDevice.DeviceID;
-                }
+                return doContinueEnumeration;
             }
+        }
+        auto devices = reinterpret_cast<std::vector<device> *>(captured_ptr);
+        device newDevice{{}, monitor};
+        if (displayDevice.DeviceID[0] != L'\0')
+        {
+            std::array<wchar_t, 256> parsedId{};
+            ParseDeviceId(displayDevice.DeviceID, parsedId.data(), size(parsedId));
+            newDevice._deviceID = parsedId.data();
+        }
+        devices->emplace_back(std::move(newDevice));
+        return doContinueEnumeration;
+    };
+    std::vector<device> devices;
+    EnumDisplayMonitors(nullptr, nullptr, callback, reinterpret_cast<LPARAM>(&devices));
+    std::vector<WmiMonitorID> monitor_infos;
+    try
+    {
+        m_wmiConnection.select_all(L"WmiMonitorID", [&](std::wstring_view xml_obj) {
+          monitor_infos.emplace_back(parse_monitorID_from_dtd(xml_obj));
+        });
+    }
+    catch(...) {}
 
-            if (validMonitor)
+    if(GetSystemMetrics(SM_REMOTESESSION))
+    {
+        // When running in a remote session, devices have "Default_Monitor#<transient_part>" DeviceIDS.
+        // Since a remote session can have multiple devices, we can't just assign them a single "REMOTE" ID.
+        // We'll assign them unique numeric IDs instead.
+        for (size_t i = 0; i < size(devices); ++i)
+        {
+            wchar_t remoteDeviceID[32];
+            swprintf_s(remoteDeviceID, L"REMOTE%zu", i);
+            AddZoneWindow(devices[i]._handle, remoteDeviceID);
+        }
+        return;
+    }
+    // If EnumDisplayMonitors returned a duplicate data, it's an API bug and we should workaround it
+    const bool enumDisplayMonitorsDublicateData = [&]{

```

Comment:
Any reason for a lambda that will only be called once?

Reply:
@bzoz to be able to mark `enumDisplayMonitorsDublicateData` as `const`! :)